contracts/Escrow.sol

```diff
@@ -0,0 +1,175 @@
+pragma solidity 0.8.9;
+
+import { IEscrow } from "./interfaces/IEscrow.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { IOracle } from "./interfaces/IOracle.sol";
+import { ILoan } from "./interfaces/ILoan.sol";
+
+contract Escrow is IEscrow {
+
+    // the minimum value of the collateral in relation to the outstanding debt e.g. 10% of outstanding debt
+    uint public minimumCollateralRatio;
+
+    // return if have collateral but no debt
+    uint256 MAX_INT = 115792089237316195423570985008687907853269984665640564039457584007913129639935;
+
+    // Stakeholders and contracts used in Escrow
+    address public loan;
+    address public oracle;
+    address public lender;
+    address public borrower;
+    address public arbiter;
+
+    // tracking tokens that were deposited
+    address[] private _tokensUsedAsCollateral;
+
+    // mapping to check uniqueness of tokensUsedAsCollateral
+    mapping(address => bool) private _tokensUsed;
```

Comment:
Three reasons: 

1. Removing a specific element will not reduce the length of the array, meaning we will still have to iterate over the full length and handle the undefined value. The length is reduced if you use `pop()` but that only removes the last element and we won't always do that
2. This method is more efficient as it will only add addresses that have never been seen and will not re add any elements to the array if they were removed prior in `releaseCollateral` or `liquidate`
3. We will not have to iterate over the entire array each time we want to remove an element because we don't need to remove any elements, we only add them and use a mapping. 

If we blindly add an address each time the deposited mapping is set to zero, it opens up an attack vector whereby the depositor can do multiple dips on the same token deposit.

Reply:
Gotcha. Do you think i should implement something similar for the `positoinIds` on the Loan contract? I put this together yesterday for managing the array we store of all currently open/active debt positions. I don't know much about arrays in solidity/EVM besides the basics so happy to move it over to a similar implementation as you've made

https://github.com/debtdao/smart-contracts/blob/780695ce2dcfbf6c637a136df29f2f1fa17fa61a/contracts/lib/LoanLib.sol#L57-L73